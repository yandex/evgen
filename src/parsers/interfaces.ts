import { groupBy, isObject } from 'lodash';

import { InterfaceVersion } from '../types/parsed-types';
import { NestedRecord, RawInterface } from '../types/raw-types';

import { VERSION_PATTERN, INTERFACES_FIELD } from './constants';
import { parseParameters } from './event-parameters';

export const parseInterfaces = (
    rawInterfaces: Record<string, NestedRecord<NestedRecord<RawInterface>>>
): Record<string, Record<string, InterfaceVersion[]>> => {
    const interfaceVersions = Object.entries(rawInterfaces).flatMap(
        ([interfaceKey, interfaceNode]) => parseInterfaceVersions([interfaceKey], interfaceNode)
    );

    const interfacesByNamespace = groupBy(
        interfaceVersions,
        (interfaceVersion) => interfaceVersion.namespace
    );

    const result: Record<string, Record<string, InterfaceVersion[]>> = {};
    Object.entries(interfacesByNamespace).forEach(([name, interfaceVersions]) => {
        result[name] = groupBy(interfaceVersions, (interfaceVersion) => interfaceVersion.name);
    });

    return result;
};

const parseInterfaceVersions = (
    namespaceParts: string[],
    node: NestedRecord<NestedRecord<RawInterface>>
): InterfaceVersion[] => {
    const namespace = namespaceParts.length > 1 ? namespaceParts[0] : INTERFACES_FIELD;
    const name = namespaceParts.length > 1 ? namespaceParts.slice(1).join('') : namespaceParts[0];
    return Object.entries(node).flatMap(([key, value]) => {
        const versionKeyMatch = key.match(VERSION_PATTERN);
        if (versionKeyMatch && versionKeyMatch[1] && isInterfaceNode(value)) {
            const version = parseInt(versionKeyMatch[1], 10);

            return [
                {
                    name,
                    namespace,
                    version,
                    parameters: parseParameters(value.parameters, name, version),
                    description: value.description || '',
                },
            ];
        }

        return isObject(value)
            ? parseInterfaceVersions(
                  [...namespaceParts, key],
                  value as NestedRecord<NestedRecord<RawInterface>>
              )
            : [];
    });
};

const isInterfaceNode = (node: NestedRecord<NestedRecord<RawInterface>>): node is RawInterface =>
    node.parameters !== undefined;
