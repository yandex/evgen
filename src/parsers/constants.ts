export const GLOBAL_PARAMETERS_FIELD = '.GlobalParams';
export const PLATFORM_PARAMETERS_FIELD = '.PlatformParams';
export const EVENTS_FIELD = 'Events';
export const INTERFACES_FIELD = 'Interfaces';
export const SHARED_FIELD = 'Shared';

export const VERSION_PATTERN = /^v(\d+)$/;
