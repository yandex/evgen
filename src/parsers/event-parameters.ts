import { ParameterType } from '../types/data-types';
import { EventParameter } from '../types/parsed-types';
import { RawEventParameter } from '../types/raw-types';

export const parseParameters = (
    parameters: Record<string, RawEventParameter> | null,
    namespace = '',
    version = 1
): EventParameter<ParameterType>[] => {
    const parsedParams: EventParameter<ParameterType>[] = [];
    if (!parameters) {
        return parsedParams;
    }
    Object.entries(parameters).forEach(([name, parameter]) => {
        parsedParams.push({
            name,
            namespace,
            version,
            type: parameter.type,
            description: parameter.description || '',
            defaultValue: parameter.default_value,
            abstract: parameter.abstract || false,
            optional: parameter.optional || false,
            elementType: parameter.element_type,
        });
    });

    return parsedParams;
};
