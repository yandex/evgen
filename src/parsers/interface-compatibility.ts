import { ParameterType } from '../types/data-types';
import { EventVersion, EventParameter, InterfaceVersion } from '../types/parsed-types';

export const getLatestCompatibleVersion = (
    eventVersion: EventVersion,
    interfaces: InterfaceVersion[],
    globalParameters: EventParameter<ParameterType>[]
) => {
    const descSortedInterfaces = interfaces.sort((i1, i2) => i2.version - i1.version);
    const latestCompatible = descSortedInterfaces.find((interfaceVersion) => {
        const compatibility = checkCompatible(eventVersion, interfaceVersion, globalParameters);
        return compatibility.isCompatible;
    });

    return latestCompatible?.version;
};

interface CompatibilityResult {
    isInProd: boolean;
    isCompatible: boolean;
    diff: EventParameter<ParameterType>[];
}

const checkCompatible = (
    eventVersion: EventVersion,
    iface: InterfaceVersion,
    globalParameters: EventParameter<ParameterType>[]
) => {
    const isInProd = eventVersion.platforms
        ? Object.values(eventVersion.platforms).some((platform) => platform.lastVersion !== null)
        : false;

    const result: CompatibilityResult = { isInProd, isCompatible: true, diff: [] };
    iface.parameters.forEach((ifaceParameter) => {
        const isCompatible =
            eventVersion.parameters.some((eventParameter) =>
                isParameterCompatible(eventParameter, ifaceParameter)
            ) ||
            globalParameters.some((globalParameter) =>
                isParameterCompatible(globalParameter, ifaceParameter)
            );
        if (!isCompatible) {
            result.isCompatible = false;
            result.diff.push(ifaceParameter);
        }
    });

    return result;
};

const isParameterCompatible = (
    param1: EventParameter<ParameterType>,
    param2: EventParameter<ParameterType>
) => param1.name === param2.name;
