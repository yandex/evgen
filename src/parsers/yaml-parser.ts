import { readFile, lstat } from 'fs/promises';
import { join } from 'path';

import { parse as yamlParse } from 'yaml';
import fg from 'fast-glob';
import { isObject, get, merge } from 'lodash';

const INCLUDE_TAG = '!include';
const INCLUDED_PREFIX = '_included';
const MERGED_PREFIX = '_merged';
const YAML_FILE_PATTERN = '**/*.yaml';

let mergedCount = 1;

const parseYamlFile = async (file: string) => {
    const yamlData = await readFile(file, { encoding: 'utf-8' });
    const replaceMergeKeys = yamlData.replaceAll(
        '<<:',
        // eslint-disable-next-line no-plusplus
        () => `${MERGED_PREFIX}_${mergedCount++}:` // use counter to make merge key unique
    );
    return yamlParse(replaceMergeKeys.replace(/\*(\S+):(\s)/g, '*$1 :$2'), {
        uniqueKeys: false,
        maxAliasCount: -1,
        customTags: [
            {
                tag: INCLUDE_TAG,
                resolve: (str: string) => `${INCLUDE_TAG}:${str}`,
            },
        ],
    });
};

type NodeType = Record<string, unknown> | string | boolean | number;

const replaceMerges = (node: unknown): unknown => {
    const merged: Record<string, unknown> = {};
    const own: Record<string, unknown> = {};

    if (isObject(node)) {
        const allKeys = Object.keys(node);
        const mergeKeys = allKeys.filter((key) => key.startsWith(MERGED_PREFIX));
        mergeKeys.forEach((key) => {
            const value = (node as Record<string, unknown>)[key];
            if (!isObject(value)) {
                throw new Error(`Included value should be object, but got ${value}`);
            }
            Object.assign(merged, replaceMerges(value));

            delete (node as Record<string, unknown>)[key];
        });

        const ownKeys = allKeys.filter((key) => !key.startsWith(MERGED_PREFIX));
        ownKeys.forEach((key) => {
            const value = (node as Record<string, unknown>)[key];
            own[key] = replaceMerges(value);

            delete (node as Record<string, unknown>)[key];
        });

        // Order according to original evgen for compatibility
        Object.assign(node, merged, own);
    }

    return node;
};

const replaceIncludes = (node: NodeType, eventsByFile: Record<string, unknown>): unknown => {
    if (typeof node === 'string' && node.startsWith(INCLUDE_TAG)) {
        const [, file, path] = node.split(':');
        const includedNode = get(eventsByFile[file], path);
        if (!includedNode) {
            throw new Error(`Failed to include ${file}: ${path}. Initial value is ${node}`);
        }

        return replaceIncludes(includedNode, eventsByFile);
    }

    if (isObject(node)) {
        const included: Record<string, unknown> = {};
        const own: Record<string, unknown> = {};

        Object.entries(node).forEach(([key, value]) => {
            const newValue = replaceIncludes(value as NodeType, eventsByFile);
            if (key.startsWith(INCLUDED_PREFIX)) {
                if (!isObject(newValue)) {
                    throw new Error(`Included value should be object, but got ${newValue}`);
                }
                Object.assign(included, newValue);
            } else {
                own[key] = newValue;
            }
            delete node[key];
        });

        // Order according to original evgen for compatibility
        return Object.assign(node, included, own);
    }

    return node;
};

export const parseConfigFile = async (filePath: string) => parseYamlFile(filePath);

export const parseEventsFile = async (eventsPath: string) => {
    const eventSource = await lstat(eventsPath);
    if (eventSource.isFile()) {
        const fileData = await parseYamlFile(eventsPath);
        return replaceMerges(fileData);
    }

    if (eventSource.isDirectory()) {
        const yamlFiles = await fg(YAML_FILE_PATTERN, { cwd: eventsPath });
        const events: Record<string, unknown> = {};
        const eventsByFile: Record<string, unknown> = {};
        await Promise.all(
            yamlFiles.map(async (fileName: string) => {
                const fileData = await parseYamlFile(join(eventsPath, fileName));
                const fileDataProcessed = replaceMerges(fileData);
                eventsByFile[fileName.replace(/\.yaml$/, '')] = fileDataProcessed;
                merge(events, fileDataProcessed);
            })
        );

        return replaceIncludes(events, eventsByFile);
    }

    throw new Error(`Invalid events path: ${eventsPath}`);
};
