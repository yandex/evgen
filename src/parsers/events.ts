import { get, isObject } from 'lodash';

import { EventNamespace, EventVersion, Event } from '../types/parsed-types';
import { NestedRecord, RawEvent } from '../types/raw-types';

import { parseParameters } from './event-parameters';
import { parsePlatforms } from './platforms';
import { VERSION_PATTERN } from './constants';
import { upperFirstLetter } from '../helpers';

const DEFAULT_NAMESPACE_NAME = 'Singletons';

export const parseEventNamespaces = (
    rawEventsNamespaces: Record<string, NestedRecord<NestedRecord<RawEvent>>>
): EventNamespace<Event>[] => {
    const docDirs: Record<string, string> = {};

    const eventVersions = Object.entries(rawEventsNamespaces).flatMap(([eventKey, eventNode]) => {
        if (get(eventNode, 'documentation_dir')) {
            docDirs[eventKey] = String(get(eventNode, 'documentation_dir'));
        }
        return parseEventVersions([eventKey], eventNode);
    });
    const eventsByNamespace = getEventsByNamespace(eventVersions);

    return Object.entries(eventsByNamespace)
        .map(([namespace, namespaceEvents]) => ({
            name: namespace,
            events: Object.entries(getEventsByName(namespaceEvents)).map(([name, versions]) => {
                if (versions.length > 1) {
                    versions.sort((v1, v2) => v1.version - v2.version);
                }
                return {
                    name,
                    versions,
                };
            }),
            documentationDir: docDirs[namespace],
        }))
        .sort((n1, n2) => {
            if (n1.name < n2.name) {
                return -1;
            }
            if (n1.name > n2.name) {
                return 1;
            }
            return 0;
        });
};

const getEventsByNamespace = (eventVersions: EventVersion[]): Record<string, EventVersion[]> =>
    eventVersions.reduce<Record<string, EventVersion[]>>((eventsByNamespace, event) => {
        const { namespace } = event;
        if (!eventsByNamespace[namespace]) {
            eventsByNamespace[namespace] = [event];
        } else {
            eventsByNamespace[namespace].push(event);
        }
        return eventsByNamespace;
    }, {});

const getEventsByName = (eventVersions: EventVersion[]): Record<string, EventVersion[]> =>
    eventVersions.reduce<Record<string, EventVersion[]>>((eventsByName, event) => {
        const { name } = event;
        if (!eventsByName[name]) {
            eventsByName[name] = [event];
        } else {
            eventsByName[name].push(event);
        }
        return eventsByName;
    }, {});

const parseEventVersions = (
    namespaceParts: string[],
    node: NestedRecord<NestedRecord<RawEvent>>
): EventVersion[] => {
    const namespace = namespaceParts.length > 0 ? namespaceParts[0] : DEFAULT_NAMESPACE_NAME;
    return Object.entries(node).flatMap(([key, value]: [string, NestedRecord<RawEvent>]) => {
        const versionKeyMatch = key.match(VERSION_PATTERN);
        if (versionKeyMatch && versionKeyMatch[1] && isEventNode(value)) {
            const name = namespaceParts.map(upperFirstLetter).join('');

            const version = parseInt(versionKeyMatch[1], 10);

            if (value.namespaces) {
                value.namespaces.forEach((namespace) => {
                    if (typeof namespace !== 'string') {
                        throw new Error(
                            `Failed to parse namespace ${JSON.stringify(
                                namespace
                            )}: namespace must be string!`
                        );
                    }
                });
            }

            return [
                {
                    name,
                    namespace,
                    event: value.force_event_name || namespaceParts.join('.'),
                    version,
                    parameters: parseParameters(value.parameters, name, version),
                    description: value.description || '',
                    comment: value.comment,
                    platforms: value.platforms ? parsePlatforms(value.platforms) : undefined,
                    interfaceNames: getInterfaceNames(value.interface),
                    interfaces: {},
                    additionalNamespaces: value.namespaces || [],
                },
            ];
        }

        if (key === 'documentation_dir' || key === 'inheritance') {
            return [];
        }

        return isObject(value)
            ? parseEventVersions(
                  [...namespaceParts, key],
                  value as NestedRecord<NestedRecord<RawEvent>>
              )
            : [];
    });
};

const getInterfaceNames = (value: unknown): string[] => {
    if (!value) {
        return [];
    }

    return Array.isArray(value) ? value : [value];
};

const isEventNode = (node: NestedRecord<NestedRecord<RawEvent>>): node is RawEvent =>
    node.parameters !== undefined && node.description !== undefined;
