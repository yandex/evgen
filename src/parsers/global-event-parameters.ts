import { isObject } from 'lodash';

import { GlobalParameters } from '../types/parsed-types';
import { RawGlobalParameters } from '../types/raw-types';

import { GLOBAL_PARAMETERS_FIELD } from './constants';
import { parseParameters } from './event-parameters';

export const parseGlobalParameters = (
    rawGlobalParameters: RawGlobalParameters
): GlobalParameters => {
    if (!rawGlobalParameters) {
        throw new Error(`Could not find definition of ${GLOBAL_PARAMETERS_FIELD}`);
    }

    if (!isObject(rawGlobalParameters)) {
        throw new Error(
            `Expected global parameters to be an object, but got ${typeof rawGlobalParameters}`
        );
    }

    const parameters = parseParameters(rawGlobalParameters.parameters);

    return {
        parameters,
        description: rawGlobalParameters.description,
        comment: rawGlobalParameters.comment,
    };
};
