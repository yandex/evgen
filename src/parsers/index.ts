import { flatten, sortBy } from 'lodash';

import { InterfaceVersion, NamespaceCollection } from '../types/parsed-types';
import { RawEvents } from '../types/raw-types';

import {
    EVENTS_FIELD,
    GLOBAL_PARAMETERS_FIELD,
    SHARED_FIELD,
    INTERFACES_FIELD,
    PLATFORM_PARAMETERS_FIELD,
} from './constants';
import { parseEventNamespaces } from './events';
import { parseGlobalParameters } from './global-event-parameters';
import { getLatestCompatibleVersion } from './interface-compatibility';
import { parseInterfaces } from './interfaces';
import { parsePlatformParameters } from './platform-parameters';

export const parseEvents = (rawEvents: RawEvents): NamespaceCollection => {
    const globalParameters =
        rawEvents[GLOBAL_PARAMETERS_FIELD] &&
        parseGlobalParameters(rawEvents[GLOBAL_PARAMETERS_FIELD]);
    const platformParameters =
        (rawEvents[PLATFORM_PARAMETERS_FIELD] &&
            parsePlatformParameters(rawEvents[PLATFORM_PARAMETERS_FIELD])) ||
        {};
    const eventNamespaces = parseEventNamespaces(rawEvents[EVENTS_FIELD]);
    const interfaceDict =
        rawEvents[INTERFACES_FIELD] && parseInterfaces(rawEvents[INTERFACES_FIELD]);

    const interfaceNamespaces: Record<string, InterfaceVersion[]> = {};

    if (interfaceDict) {
        sortBy(Object.entries(interfaceDict), (item) => item[0]) // sort by namespace for order in docs
            .forEach(([interfaceNamespace, events]) => {
                interfaceNamespaces[interfaceNamespace] = flatten(Object.values(events));
            });

        eventNamespaces.forEach((namespace) =>
            namespace.events.forEach((event) =>
                event.versions.forEach((eventVersion) => {
                    if (eventVersion.interfaceNames?.length) {
                        const interfaces: Record<string, number> = {};
                        eventVersion.interfaceNames.forEach((interfaceFullName) => {
                            const interfaceParts = interfaceFullName.split('.', 2);
                            const [namespace, name] =
                                interfaceParts.length > 1
                                    ? interfaceParts
                                    : [INTERFACES_FIELD, interfaceParts[0]];
                            if (!interfaceDict[namespace] || !interfaceDict[namespace][name]) {
                                console.log(`Interface ${interfaceFullName} not found.`);
                                return;
                            }
                            const latestVersion = getLatestCompatibleVersion(
                                eventVersion,
                                interfaceDict[namespace][name],
                                globalParameters.parameters
                            );
                            if (latestVersion) {
                                interfaces[interfaceFullName] = latestVersion;
                            } else {
                                console.log(
                                    `Event ${eventVersion.namespace}:${eventVersion.name} is not compatible with ${interfaceFullName}`
                                );
                            }
                        });

                        eventVersion.interfaces = interfaces;
                    }
                })
            )
        );
    }
    return {
        globalParameters,
        platformParameters,
        eventNamespaces,
        interfaceNamespaces,
        shared: rawEvents[SHARED_FIELD],
    };
};
