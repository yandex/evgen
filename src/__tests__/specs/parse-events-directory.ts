import { parseEventsFile } from '../../parsers/yaml-parser';

describe('parse event yaml directory', () => {
    it('parse events directory yaml with includes', async () => {
        expect(await parseEventsFile('src/__tests__/examples/events')).toMatchSnapshot();
    });
});
