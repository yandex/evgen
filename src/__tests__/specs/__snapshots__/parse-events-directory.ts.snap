// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse event yaml directory parse events directory yaml with includes 1`] = `
Object {
  ".GlobalParams": Object {
    "description": "Эти параметры добавляются к параметрам каждого события.",
    "parameters": Object {
      "globalParam": Object {
        "description": "Глобальный параметр, который добавится к каждомоу событию",
        "type": "String",
      },
    },
  },
  "Events": Object {
    "AnotherNamespace": Object {
      "Event1": Object {
        "v1": Object {
          "description": "Первое событие с переиспользуемым параметром",
          "parameters": Object {
            "reusedParam": Object {
              "description": "Параметр, который переиспользуется в нескольких событиях",
              "type": "String",
            },
          },
          "platforms": Object {
            "Android": Object {
              "app_versions": "in_progress",
              "ticket": "https://your-tracker.com",
            },
            "WebSmartTV": Object {
              "app_versions": "in_progress",
              "ticket": "https://your-tracker.com",
            },
            "iOS": Object {
              "app_versions": "in_progress",
              "ticket": "https://your-tracker.com",
            },
          },
        },
      },
      "Event2": Object {
        "v1": Object {
          "description": "Второе событие с переиспользуемым параметром",
          "parameters": Object {
            "reusedParam": Object {
              "description": "Параметр, который переиспользуется в нескольких событиях",
              "type": "String",
            },
          },
          "platforms": Object {
            "Android": Object {
              "app_versions": "in_progress",
              "ticket": "https://your-tracker.com",
            },
            "WebSmartTV": Object {
              "app_versions": "in_progress",
              "ticket": "https://your-tracker.com",
            },
            "iOS": Object {
              "app_versions": "in_progress",
              "ticket": "https://your-tracker.com",
            },
          },
        },
      },
      "documentation_dir": "AnotherDir",
      "inheritance": Object {
        "reused_param": Object {
          "reusedParam": Object {
            "description": "Параметр, который переиспользуется в нескольких событиях",
            "type": "String",
          },
        },
      },
    },
    "MyNamespace": Object {
      "MyEvent": Object {
        "v1": Object {
          "description": "События со всеми возможными типами параметров",
          "parameters": Object {
            "batchParam1": Object {
              "description": "Параметр, описанный в отдельным файле.",
              "type": "String",
            },
            "batchParam2": Object {
              "description": "Параметр, описанный в отдельным файле.",
              "type": "String",
            },
            "intParam": Object {
              "default_value": 42,
              "description": "Параметр типа Int",
              "type": "Int",
            },
            "paramFromAnotherFile": Object {
              "description": "Параметр, описанный в отдельным файле.",
              "type": "String",
            },
            "stringParam": Object {
              "default_value": "val1",
              "description": "Парамтер типа String",
              "type": "String",
            },
            "сonstParam": Object {
              "description": "Constant parameter",
              "type": Object {
                "Const": "shop_page",
              },
            },
          },
          "platforms": Object {
            "Android": Object {
              "app_versions": 3.14,
              "ticket": "https://your-tracker.com",
            },
            "WebSmartTV": Object {
              "app_versions": "in_progress",
              "ticket": "https://your-tracker.com",
            },
            "iOS": Object {
              "app_versions": 4.13,
              "ticket": "https://your-tracker.com",
            },
          },
        },
      },
    },
  },
  "reused_params": Object {
    "several_params": Object {
      "batchParam1": Object {
        "description": "Параметр, описанный в отдельным файле.",
        "type": "String",
      },
      "batchParam2": Object {
        "description": "Параметр, описанный в отдельным файле.",
        "type": "String",
      },
    },
    "single_param": Object {
      "paramFromAnotherFile": Object {
        "description": "Параметр, описанный в отдельным файле.",
        "type": "String",
      },
    },
  },
  "reused_scalar_values": Object {
    "shop_page": "shop_page",
  },
}
`;
